src/transport/SessionManager.cpp

```diff
@@ -668,68 +668,53 @@ void SessionManager::OnMessageReceived(const PeerAddress & peerAddress, System::
 {
     char peerAddrBuf[chip::Transport::PeerAddress::kMaxToStringSize];
 
-    VerifyOrReturn(conn != nullptr);
-    conn->mPeerAddr.ToString(peerAddrBuf);
+    conn.mPeerAddr.ToString(peerAddrBuf);
     ChipLogProgress(Inet, "Received TCP connection request from %s.", peerAddrBuf);
 
-    Transport::AppTCPConnectionCallbackCtxt * appTCPConnCbCtxt = conn->mAppState;
+    Transport::AppTCPConnectionCallbackCtxt * appTCPConnCbCtxt = conn.mAppState;
     if (appTCPConnCbCtxt != nullptr && appTCPConnCbCtxt->connReceivedCb != nullptr)
     {
         appTCPConnCbCtxt->connReceivedCb(conn);
     }
 }
 
-void SessionManager::HandleConnectionAttemptComplete(Transport::ActiveTCPConnectionState * conn, CHIP_ERROR conErr)
+void SessionManager::HandleConnectionAttemptComplete(Transport::ActiveTCPConnectionHolder & conn, CHIP_ERROR conErr)
 {
-    VerifyOrReturn(conn != nullptr);
+    VerifyOrReturn(!conn.IsNull());
 
     Transport::AppTCPConnectionCallbackCtxt * appTCPConnCbCtxt = conn->mAppState;
-    if (appTCPConnCbCtxt == nullptr)
-    {
-        TCPDisconnect(conn, /* shouldAbort = */ true);
-        return;
-    }
-
-    if (appTCPConnCbCtxt->connCompleteCb != nullptr)
+    bool callbackHandled                                       = false;
+    if (appTCPConnCbCtxt != nullptr && appTCPConnCbCtxt->connCompleteCb != nullptr)
     {
         appTCPConnCbCtxt->connCompleteCb(conn, conErr);
+        callbackHandled = true;
     }
-    else
+
+    if ((mConnDelegate == nullptr || !mConnDelegate->OnTCPConnectionAttemptComplete(conn, conErr)) && !callbackHandled)
     {
         char peerAddrBuf[chip::Transport::PeerAddress::kMaxToStringSize];
         conn->mPeerAddr.ToString(peerAddrBuf);
 
-        ChipLogProgress(Inet, "TCP Connection established with peer %s, but no registered handler. Disconnecting.", peerAddrBuf);
-
-        // Close the connection
-        TCPDisconnect(conn, /* shouldAbort = */ true);
+        ChipLogProgress(Inet, "TCP Connection established with peer %s, but no registered handler.", peerAddrBuf);
     }
 }
 
-void SessionManager::HandleConnectionClosed(Transport::ActiveTCPConnectionState * conn, CHIP_ERROR conErr)
+void SessionManager::HandleConnectionClosed(Transport::ActiveTCPConnectionState & conn, CHIP_ERROR conErr)
 {
-    VerifyOrReturn(conn != nullptr);
-
-    MarkSecureSessionOverTCPForEviction(conn, conErr);
-
-    // TODO: A mechanism to mark an unauthenticated session as unusable when
-    // the underlying connection is broken. Issue #32323
-
-    Transport::AppTCPConnectionCallbackCtxt * appTCPConnCbCtxt = conn->mAppState;
-    VerifyOrReturn(appTCPConnCbCtxt != nullptr);
-
-    if (appTCPConnCbCtxt->connClosedCb != nullptr)
+    Transport::AppTCPConnectionCallbackCtxt * appTCPConnCbCtxt = conn.mAppState;
+    if (appTCPConnCbCtxt != nullptr && appTCPConnCbCtxt->connClosedCb != nullptr)
     {
         appTCPConnCbCtxt->connClosedCb(conn, conErr);
     }
+    MarkSecureSessionOverTCPForEviction(conn, conErr);
```

Comment:
Why the ordering change here?  If the order is important, please document why it has to be this order.